### PR TITLE
Add model STSSL.

### PR DESCRIPTION
**Pull request by Lucas-lyh**
该PR添加了STSSL模型，使用STSSL论文代码添加了model，添加了自适应loss权重的executor，添加了IN/OUT+MAE/MAPE的评估器以复现结果。

复现结果如下，标黄为复现指标
bc37e2d3c17436f195dd35a556538e9

### CODE REVIEW

File: libcity/model/loss.py [hunk 1]
```diff
@@ -71,6 +87,23 @@ def masked_mape_torch(preds, labels, null_val=np.nan, eps=0):
     loss = torch.where(torch.isnan(loss), torch.zeros_like(loss), loss)
     return torch.mean(loss)
 
+def masked_mape_torch_with_mask_value(preds, labels, null_val=np.nan, eps=0, mask_value=5):
```
Comment (Kazeya27) [inline libcity/model/loss.py:90]: 这里不用重新写一个方法，可以改一些原来的loss方法，添加一个mask_value参数，然后默认为None，不为None的时候再去mask

File: libcity/evaluator/traffic_state_in_out_evaluator.py [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+import datetime
+import pandas as pd
+from libcity.utils import ensure_dir
+from libcity.model import loss
+from logging import getLogger
+from libcity.evaluator.abstract_evaluator import AbstractEvaluator
+from libcity.evaluator.traffic_state_evaluator import TrafficStateEvaluator
+
+
+class TrafficStateInOutEvaluator(TrafficStateEvaluator):
```
Comment (Kazeya27) [inline libcity/evaluator/traffic_state_in_out_evaluator.py:12]: 同样的，改写loss函数后，直接在原来TrafficStateEvaluator的基础上改改就行，不需要重写一个类

File: libcity/data/dataset/dataset_subclass/stssl_dataset.py [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+    def get_data_feature(self):
+        """
+        返回数据集特征，scaler是归一化方法，adj_mx是邻接矩阵，num_nodes是网格的个数，
+        len_row是网格的行数，len_column是网格的列数，
+        feature_dim是输入数据的维度，output_dim是模型输出的维度
+
+        Returns:
+            dict: 包含数据集的相关特征的字典
+        """
+        lp = self.len_period * (self.pad_forward_period + self.pad_back_period + 1)
```
Comment (Kazeya27) [inline libcity/data/dataset/dataset_subclass/stssl_dataset.py:31]: 如果只是简单的返回这两个值，是不是在model类里面从config获取就行，不用再写一个dataset类

File: libcity/model/traffic_flow_prediction/STSSL.py [hunk 4]
```diff
@@ -0,0 +1,647 @@
...
+        sep_loss.append(l3.item())
+        loss += loss_weights[2] * l3
+        return loss, sep_loss
+
+    def pred_loss(self, z1, z2, y_true, scaler):
+        y_pred = scaler.inverse_transform(self._predict(z1, z2))
+        y_true = scaler.inverse_transform(y_true)
+
+        loss = self.yita * self.mae(y_pred[..., 0], y_true[..., 0]) + \
+               (1 - self.yita) * self.mae(y_pred[..., 1], y_true[..., 1])
```
Comment (Kazeya27) [inline libcity/model/traffic_flow_prediction/STSSL.py:623]: 为什么不直接用loss.py里的函数呢？

**Review by Kazeya27 — CHANGES_REQUESTED** [review]

File: libcity/model/loss.py [hunk 5]
```diff
@@ -14,12 +14,16 @@ def masked_mae_loss(y_pred, y_true):
...
+def masked_mae_torch(preds, labels, null_val=np.nan, reduce=True, mask_val=np.nan, eval_mask=None):
     labels[torch.abs(labels) < 1e-4] = 0
     if np.isnan(null_val):
         mask = ~torch.isnan(labels)
     else:
         mask = labels.ne(null_val)
+    if not np.isnan(mask_val):
+        mask &= labels.ge(mask_val)
+    if eval_mask is not None:
+        mask &= eval_mask
```
Comment (Kazeya27) [inline libcity/model/loss.py:26]: eval_mask和你的实验没有关系的，不用加

**Review by Kazeya27 — CHANGES_REQUESTED** [review]

**Comment by Lucas-lyh** [comment]
已修复，去除了flow预测特有的in/out分维度评测，使用通用的评测指标。

File: libcity/evaluator/traffic_state_evaluator.py [hunk 6]
```diff
@@ -12,10 +12,11 @@ class TrafficStateEvaluator(AbstractEvaluator):
 
     def __init__(self, config):
         self.metrics = config.get('metrics', ['MAE'])  # 评估指标, 是一个 list
-        self.allowed_metrics = ['MAE', 'MSE', 'RMSE', 'MAPE', 'masked_MAE',
-                                'masked_MSE', 'masked_RMSE', 'masked_MAPE', 'R2', 'EVAR']
+        self.allowed_metrics = ['MAE', 'MSE', 'RMSE', 'MAPE', 'masked_MAE', 'masked_MSE', 'masked_RMSE', 'masked_MAPE',
+                                'R2', 'EVAR', "IN_masked_MAE", "IN_masked_MAPE", "OUT_masked_MAE", "OUT_masked_MAPE"]
```
Comment (Kazeya27) [inline libcity/evaluator/traffic_state_evaluator.py:16]: 这里也要删呀

File: libcity/evaluator/traffic_state_evaluator.py [hunk 7]
```diff
@@ -12,10 +12,11 @@ class TrafficStateEvaluator(AbstractEvaluator):
...
         self.save_modes = config.get('save_mode', ['csv', 'json'])
         self.mode = config.get('evaluator_mode', 'single')  # or average
+        self.mask_val = config.get('eval_mask_val', None)
```
Comment (Kazeya27) [inline libcity/evaluator/traffic_state_evaluator.py:19]: 直接叫mask_val就可以

File: libcity/evaluator/traffic_state_evaluator.py [hunk 8]
```diff
@@ -88,34 +91,36 @@ def collect(self, batch):
                 for metric in self.metrics:
                     if metric == 'masked_MAE':
                         self.intermediate_result[metric + '@' + str(i)].append(
-                            loss.masked_mae_torch(y_pred[:, i-1], y_true[:, i-1], 0).item())
+                            loss.masked_mae_torch(y_pred[:, i - 1], y_true[:, i - 1], 0,
+                                                  mask_val=self.mask_val).item())
                     elif metric == 'masked_MSE':
                         self.intermediate_result[metric + '@' + str(i)].append(
-                            loss.masked_mse_torch(y_pred[:, i-1], y_true[:, i-1], 0).item())
+                            loss.masked_mse_torch(y_pred[:, i - 1], y_true[:, i - 1], 0).item())
```
Comment (Kazeya27) [inline libcity/evaluator/traffic_state_evaluator.py:98]: 把MSE和RMSE也加上这个参数吧

File: libcity/model/loss.py [hunk 9]
```diff
@@ -54,7 +56,7 @@ def quantile_loss(preds, labels, delta=0.25):
     return torch.mean(torch.where(condition, large_res, small_res))
 
 
-def masked_mape_torch(preds, labels, null_val=np.nan, eps=0):
+def masked_mape_torch(preds, labels, null_val=np.nan, eps=0, mask_val=np.nan):
```
Comment (Kazeya27) [inline libcity/model/loss.py:59]: 这里是不是搞忘改了、、

**Review by Kazeya27 — CHANGES_REQUESTED** [review]

**Review by Kazeya27 — APPROVED** [review]

**Review by Kazeya27 — APPROVED** [review]
